### PR TITLE
Add support for generic types with JsonQualifiers (#92)

### PR DESCRIPTION
**Pull request by mattbdean**
Generated JsonAdapters now supports generic properties that use JsonQualifiers.

I'm kinda new to AutoValue extensions, so please let me know if anything could be written better/more logically!

### CODE REVIEW

**Review by ZacSweers — APPROVED** [review]
This looks good to me. Going to see if I can get another set of eyes on this approach before merging. Thanks for the PR!

File: auto-value-moshi/src/main/java/com/ryanharter/auto/value/moshi/AutoValueMoshiExtension.java
Comment (ZacSweers) [inline auto-value-moshi/src/main/java/com/ryanharter/auto/value/moshi/AutoValueMoshiExtension.java]: Does this assume there's only one type parameter?

**Review by ZacSweers — CHANGES_REQUESTED** [review]
Added a comment for a possible issue (requesting changes on that).

Thinking more, the reflective bit for getting the annotations makes sense, but it's also not ideal since it breaks proguard and also has reflection cost. One idea is to create a proxy of the target annotation, but maybe that's out of scope for this RFC. What do you think?

An example of what I'm thinking: https://github.com/square/moshi/blob/master/moshi/src/main/java/com/squareup/moshi/Types.java#L173-L203

File: auto-value-moshi/src/main/java/com/ryanharter/auto/value/moshi/AutoValueMoshiExtension.java
Comment (mattbdean) [reply auto-value-moshi/src/main/java/com/ryanharter/auto/value/moshi/AutoValueMoshiExtension.java]: That is correct. The idea was just to introduce the feature, so right now a property with the type `Map<T, U>` will not work as intended. I can take another look at it if this is a big issue.

**Comment by mattbdean** [comment]
Do we have a rough estimate on how much faster it'll be? If it's just a matter of a few nanoseconds I wouldn't worry about it. On the other hand, we should consider the extreme of a data class with 1000's of properties (however unrealistic that is).

Conceptually, it sounds like a very good idea. It does seem like it'll increase the complexity/readability of the generated code. If it gets implemented it definitely deserves its own PR.

File: auto-value-moshi/src/main/java/com/ryanharter/auto/value/moshi/AutoValueMoshiExtension.java
Comment (ZacSweers) [reply auto-value-moshi/src/main/java/com/ryanharter/auto/value/moshi/AutoValueMoshiExtension.java]: yeah if you wouldn't mind. I don't think we want to only have partial support, and I think it should be doable to just match on all the type arguments

**Comment by ZacSweers** [comment]
So speed isn't really the main bit for me, rather that what this library offers is actually a route to escape needing to do custom rules for reflection. Agreed that should be a separate change though, since we have existing adapter code already anyway doing this

**Comment by mattbdean** [comment]
Basically what I've done in this last commit is modified the `makeType` method so that it will work with generic types as well. Since `makeType` was already capable of using the Types class for complex types like `List<Map<String, List<Integer>>>`, generic types should now be fully supported with this PR. In other words, we can now do something like this:

```java
@AutoValue
public abstract class Foo<T, U> {
    public abstract Map<U, T> bar();
    public abstract List<T> baz();
    public abstract U qux();

    // <adapter factory method>
}
```

**Comment by ZacSweers** [comment]
This looks reasonable to me. @rharter do you want to take a look too?

**Review by ZacSweers — APPROVED** [review]

**Comment by rharter** [comment]
This looks good to me.

It looks like users are going to need to put `@Keep` annotations on their simple generic types or add rules if they use proguard, since those are fetched by name.  I'm not sure of the best way to handle that, perhaps just a note in the readme?

**Review by rharter — APPROVED** [review]

**Comment by mattbdean** [comment]
Is there anything else we need to do before this gets merged?

**Comment by ZacSweers** [comment]
Oops just saw Ryan's comment! I think just a note in the README should suffice. We could explore possibly adding them ourselves if they're present, but worth a separate PR I think and possibly avoidable if we used something like `AutoAnnotation`

**Comment by ZacSweers** [comment]
Thanks @mattbdean!

**Pull request merged**
